### ISSUE TITLE
Issue with paths between dev and staging

### ISSUE THREAD

**Issue by bossanova808** [aelvan/Imager-Craft#17]
Ok, so as per slack I am having some paths/duplicate file type issues and imager used with Amazon S3.  It's a little confusing as I am not 100% sure what is _supposed_ to be happening.  I assumed all images created by any of my environments would be push to S3 and then appear at the same path on S3 (as happens with native transforms) - but this doesn't seem to be happening with imager.

They result is that I must first load a page on my dev server before the images will load on my staging server (otherwise I get image not found errors).  T

My settings are currently:
```
  'imagerSystemPath' => $_SERVER['DOCUMENT_ROOT'] . '/imager-transforms/',
  'imagerUrl' => 'https://images.imagescience.com.au.s3-ap-southeast-2.amazonaws.com/imager-transforms/',
```
(For all hosts - not sure if I should be setting a different imagerUrl for staging, or if that is possible?

Dev (local) host - ubuntu running LAMP stack
Staging host - Vultr running a droplet managed by ServerPilot

My dev process involves pushing my latest changes and db out to staging regularly, and my guess is there's some reference to the  image paths which then doesn't work on staging, but here's one example of what seems to be the issue:

I load a page on dev, and some images are thus resized.  They then appear here

On dev:
 All Buckets /images.imagescience.com.au/imager-transforms/images_imagescience_com_au/Products/Canson/Canvas

...and that's the path I would have expected all environments to push the images to.

On staging, however, it seems images are created and then pushed to:
 All Buckets /images.imagescience.com.au/imager-transforms/srv/users/serverpilot/apps/imagescience/releases/dev/public_html/imager-transforms/images_imagescience_com_au/Products/Canson/Canvas

If I manually check the urls of the images on staging, they look like this:
https://images.imagescience.com.au.s3-ap-southeast-2.amazonaws.com/imager-transforms/images_imagescience_com_au/Products/X-Rite/Display/318616df8ae6787835eb0a6e3f3a541b.jpg

...but that is the wrong url, as the images are actually being created at that much longer url above it seems.  Which means image not found, of course.

Hoping you can help (once back from your holiday!), I have a big Commerce site with a LOT images that I am trying to make look really nice.  Let me finish by saying I love you work here - being able to do things like use an unsharp mask on downsizing large images has a *substantial* effect on the resulting quality of images, and is very awesome.  Many thanks for imager!

**Comment by bossanova808** [comment]
Just an update, I have changed settings to remove the specifc s3 region part, so now it's just:
```
  'imagerSystemPath' => $_SERVER['DOCUMENT_ROOT'] . '/imager-transforms/',
  'imagerUrl' => 'https://images.imagescience.com.au/imager-transforms/',
```
(because otherwise I get certificate errors about the domain name not matching when serving over https, which I do for everything, via cloudflare).  

But that doesn't effect the paths issue described, I'm just letting you know for completeness.

**Comment by aelvan** [comment]
Ok, first of all, when testing, make sure that you empty cache on your staging server, just to rule out that any settings from your dev environment has been cached. But yeah, as long as you have the same setup both on dev and staging, loading and transforming images to S3, this should work the same.

Make sure you're using the latest version of Imager, there were some changes to how AWS-uploading works in the latest commits.

Even though you've chosen to upload images to S3, images will still be created in the `imagerSystemPath` in your dev and staging environments. Take a look in that path on your staging server, are the transformed images created in the correct folder (after emptying the imager cache), or is the document root path (the /srv/users/serverpilot/apps/....etc... part above) also shown in there?

Also, have you set the `awsFolder` config setting to `imager-transforms`? I assume that you have since it works in dev, just wanted to make sure. By default, the files are uploaded to the root of the S3 bucket, the last part of the `imagerSystemPath` isn't automatically added to the upload path.

If all this is ok, I'm suspecting that this has something to do with the path setup on your server. Are you using some kind of linking to put the latest deployed version into production? Could you try creating a php file on the staging server with the following code in it, and sending me the result either here or on slack:

```
<?php 
   echo $_SERVER['DOCUMENT_ROOT'];
   echo '<br>';
   echo realpath($_SERVER['DOCUMENT_ROOT']);
```

**Comment by aelvan** [comment]
Disregard my comment about the awsFolder setting, this is of course correct if that path is the url to your bucket.

**Comment by bossanova808** [comment]
(Ok I am pretty sure all the config is fine and I have been clearing caches etc.)

On dev and staging, the local transforms both appear under
`craft/public_html/imager-transforms`
as I expect, with the same paths.

However dev and staging are definitely creating files in different subfolders on S3:
dev - directly under `bucket/imager-transforms`
staging - `bucket//srv/users/serverpilot/apps/imagescience/releases/dev/public_html/imager-transforms`

I think you've hit the nail on the head with your last paragraph - I am indeed doing releases with symlinks on that server (basically I can auto deploy from a date, a git tag, or just current dev code via ansible - see https://github.com/bossanova808/CraftAnsible if you're interested) 

The result of your script is:
```
/srv/users/serverpilot/apps/imagescience/public
/srv/users/serverpilot/apps/imagescience/releases/dev/public_html
```
So basically yeah, it's the link bit inbetween that shouldn't be there - `/releases/dev/public_html/` - but not sure what best solution is?

Thanks again for all your help, very much appreciated!

**Comment by aelvan** [comment]
In your config, try using:

'imagerSystemPath' => realpath($_SERVER['DOCUMENT_ROOT']) . '/imager-transforms/',

**Comment by bossanova808** [comment]
For both servers or just staging?

**Comment by aelvan** [comment]
Yes, it shouldn't make a difference on dev, what we want is the real path to the public folder on the staging server.

**Comment by bossanova808** [comment]
ok will give it a try in the morning - thanks!

**Comment by bossanova808** [comment]
Yep this seems to work fine in initial testing and it's creating all the images under the same folder on S3 irregardless of whether first transformed on dev or staging.  Great!

Given Serverpilot setups are very popular with Craft people, might be worth a note in the docs perhaps?

Thanks once again for your awesome help :) !
